# Best Model Evaluation: Cold start EfficientNetB7-ResidualMLP Neural Architecture Search using hyperband (on CIFAR10 dataset)
## Ran task.py with default settings on 2022-02-04
## Here, we evaluate the best model from 12 hours of training / search
## The first several cells are re-creating the environment from the run. The first 2 are displaying the NAS task (task.py) which was run and the content of my Python package residualmlp that was used:

In [3]:
! cat task.py

import os
import subprocess
import argparse
# Add parser for common params

try:
    import pendulum
except ModuleNotFoundError as not_found:
    print("We're missing a module:")
    print(not_found)
    print("No problem, we install it...")
    subprocess.run("pip3 install pendulum", 
    	           shell=True, 
    	           check=True)
    import pendulum

try:
    import pandas as pd
except ModuleNotFoundError as not_found:
    print("We're missing a module:")
    print(not_found)
    print("No problem, we install it...")
    subprocess.run("pip3 install pandas", 
    	           shell=True, 
    	           check=True)
    import pandas as pd

try:
    import keras_tuner as kt
except ModuleNotFoundError as not_found:
    print("We're missing a module:")
    print(not_found)
    print("No problem, we install it...")
    subprocess.run("pip3 install keras_tuner", 
    	           shell=True, 
    	           check=True)
    import keras_tuner as kt


try:
    import tensorflow as

In [4]:
! cat residualmlp/residual_mlp.py

try:
    import keras_tuner as kt
except Exception as exc:
    print("Importing Keras tuner appears to be unsuccesful. "
          "keras tuner may need to be installed $ pip install -q -U "
          "keras-tuner. The auto-ml features are disabled until this is "
           "fixed, but ResidualMLP will work. A more detailed error is: "
           f"{exc}")
import tensorflow as tf
import numpy as np
import pandas as pd
import pendulum

# Becomes a layer to convert BW images to RGB

class ResidualMLP:
    
    def __init__(self, problem_type = 'classification',
                      learning_rate = .0007,
                      minimum_learning_rate = 0.00007,
                      maximum_learning_rate = 0.7,
                      number_of_learning_rates_to_try = 5,
                      input_shape = (32,32,3),
                      bw_images = False,
                      base_model = '',
                      base_model_input_shape = (600,600,3),
                      base_model_hyp

In [5]:
import os
import subprocess
import argparse
# Add parser for common params

try:
    import pendulum
except ModuleNotFoundError as not_found:
    print("We're missing a module:")
    print(not_found)
    print("No problem, we install it...")
    subprocess.run("pip3 install pendulum", 
    	           shell=True, 
    	           check=True)
    import pendulum

try:
    import pandas as pd
except ModuleNotFoundError as not_found:
    print("We're missing a module:")
    print(not_found)
    print("No problem, we install it...")
    subprocess.run("pip3 install pandas", 
    	           shell=True, 
    	           check=True)
    import pandas as pd

try:
    import keras_tuner as kt
except ModuleNotFoundError as not_found:
    print("We're missing a module:")
    print(not_found)
    print("No problem, we install it...")
    subprocess.run("pip3 install keras_tuner", 
    	           shell=True, 
    	           check=True)
    import keras_tuner as kt


try:
    import tensorflow as tf
except ModuleNotFoundError as not_found:
    print("We're missing a module:")
    print(not_found)
    print("No problem, we install it...")
    subprocess.run("pip3 install tensorflow", 
    	           shell=True, 
    	           check=True)
    import tensorflow as tf

from residualmlp.residual_mlp import ResidualMLP

We're missing a module:
No module named 'pendulum'
No problem, we install it...


You should consider upgrading via the '/usr/bin/python -m pip install --upgrade pip' command.


We're missing a module:
No module named 'pandas'
No problem, we install it...


You should consider upgrading via the '/usr/bin/python -m pip install --upgrade pip' command.


We're missing a module:
No module named 'keras_tuner'
No problem, we install it...


You should consider upgrading via the '/usr/bin/python -m pip install --upgrade pip' command.


In [6]:
cifar = tf.keras.datasets.cifar10.load_data()
(x_train, y_train), (x_test, y_test) = cifar

y_train_ohe = tf.one_hot([i[0] for i in  y_train],10)
indexes_for_rows = tf.range(0,y_train.shape[0])
shuffled_indexes = tf.random.shuffle(indexes_for_rows)
# selected_indexes = shuffled_indexes[:TRAINING_SET_SIZE]
# selected_x_train = x_train[selected_indexes,:,:,:]
# selected_y_train_ohe = y_train_ohe.numpy()[selected_indexes,:]


170508288/170498071 [==============================] - 8s 0us/step


2022-02-05 06:30:11.205215: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1050] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2022-02-05 06:30:11.213290: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1050] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2022-02-05 06:30:11.213907: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1050] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2022-02-05 06:30:11.215776: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1050] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2022-02-05 06:30:11.216425: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1050] successful NUMA node read f

In [7]:
y_test_ohe = tf.one_hot([i[0] for i in y_test],10)

### Below is copied and pasted from task.py to re-instantiate the default settings used for the original hyperband NAS run when task.py was run. (... even though we are not running this as a script and it otherwise doesn't make sense why I have argparse content here... It is just  a shortcut to get te same global settings back.)

In [8]:

parser = argparse.ArgumentParser()

parser.add_argument(
    "--project_name", type=str, 
    help="Name for this project.",
    default="CIFAR10_EfficientNetB7-ResidualMLP_NAS_AUGMENTED_SPACE")

parser.add_argument(
    "--training_set_size", 
    type=int,
    help="Training set size (how many observations).",
    default=50000)

parser.add_argument(
    "--patience",
    type=int,
    help="How many epochs with no improved performance before the "
        "early stopping callback stops further training?",
    default=25)

parser.add_argument(
    "--patience_min_delta",
    type=float,
    help="How sensitive should the early stopping callback be"
        "  to change?",
    default=0.00001)

parser.add_argument(
    "--batch_size",
    type=int,
    help="How many observations to train with...",
    default=50)

parser.add_argument(
    "--max_epochs",
    type=int,
    help="max_epochs: Integer, the maximum number of epochs to train one "
         "model. It is recommended to set this to a value slightly higher "
         "than the expected epochs to convergence for your largest Model, "
         "and to use early stopping during training (for example, via .",
         default=50)

parser.add_argument(
        "--minimum_learning_rate",
        type=float,
        help="Lowest learning rate to try?",
        default = 0.00007)

parser.add_argument(
        "--maximum_learning_rate",
        type=float,
        help="Highest learning rate to try?",
        default = 0.7)

parser.add_argument(
        "--number_of_learning_rates_to_try",
        type=int,
        help="How many learning rate to try (maximum)?",
        default = 7)

parser.add_argument(
        "--minimum_number_of_blocks",
        type=int,
        help="Minimum number of ResidualMLP blocks in neural architectures "
             "to try?",
        default = 1)

parser.add_argument(
        "--maximum_number_of_blocks",
        type=int,
        help="Maximum number of ResidualMLP blocks in neural architectures "
              "to try?",
        default = 8)


parser.add_argument(
        "--minimum_number_of_layers_per_block",
        type=int,
        help="Minimum number of layers to try in each ResidualMLP block in "
             "the neural architectures to try?",
        default = 1)

parser.add_argument(
        "--maximum_number_of_layers_per_block",
        type=int,
        help="Maximum number of layers to try in each ResidualMLP block in "
             "the neural architectures to try?",
        default = 8)

parser.add_argument(
        "--minimum_neurons_per_block_layer",
        type=int,
        help="Minimum number of neurons to try in the Dense layers in "
             "each ResidualMLP block in the neural architectures to tried?",
        default = 30)

parser.add_argument(
        "--maximum_neurons_per_block_layer",
        type=int,
        help="Maximum number of neurons to try in the Dense layers in "
             "each ResidualMLP block in the neural architectures to tried?",
        default = 130)

parser.add_argument(
        "--n_options_of_neurons_per_layer_to_try",
        type=int,
        help="How many different numbers of neurons (at most) to try in "
             "the Dense layers try in each ResidualMLP block in the neural "
             "architectures to tried?",
        default = 7)

parser.add_argument(
        "--minimum_neurons_per_block_layer_decay",
        type=int,
        help="Lowest decay in number of neurons (n less neurons than the "
             "last layer) to try in the Dense layers try in each "
             "ResidualMLP block in the neural "
             "architectures to tried?",
        default = 0)

parser.add_argument(
        "--maximum_neurons_per_block_layer_decay",
        type=int,
        help="Highest decay in number of neurons (n less neurons than the "
             "last layer) to try in the Dense layers try in each "
             "ResidualMLP block in the neural "
             "architectures to tried?",
        default = 50)

parser.add_argument(
        "--minimum_dropout_rate_for_bypass_layers",
        type=float,
        help="Lowest dropout rate to try for dropout layers located in "
             "residual byass layers?",
        default = 0.01)

parser.add_argument(
        "--maximim_dropout_rate_for_bypass_layers",
        type=float,
        help="Highest dropout rate to try for dropout layers located in "
             "residual byass layers?",
        default = 0.7)

parser.add_argument(
        "--n_options_dropout_rate_for_bypass_layers",
        type=int,
        help="How many dropout rates to try for dropout layers located in "
             "residual byass layers?",
        default = 7)

parser.add_argument(
        "--minimum_inter_block_layers_per_block",
        type=int,
        help="Minimum number of neurons per Dense layer for Dense layers "
             "inserted between ResidualMLP blocks? 0 Will not insert a "
             "layer if selected.",
        default = 0)

parser.add_argument(
        "--maximum_inter_block_layers_per_block",
        type=int,
        help="Maximum number of neurons per Dense layer for Dense layers "
             "inserted between ResidualMLP blocks? 0 Will not insert a "
             "layer if selected.",
        default = 150)

parser.add_argument(
        "--n_options_inter_block_layers_per_block",
        type=int,
        help="How many different numbers of neurons per Dense layer for "
             "Dense layers inserted between ResidualMLP blocks will we try "
             "(at most)?",
        default = 7)

parser.add_argument(
        "--minimum_dropout_rate",
        type=float,
        help="Minimum dropout rate for final Dense layers?",
        default = 0.01)

parser.add_argument(
        "--maximum_dropout_rate",
        type=float,
        help="Maximum dropout rate for final Dense layers?",
        default = 0.7)

parser.add_argument(
        "--n_options_dropout_rate",
        type=int,
        help="How many dropout rates (at max) to try for final Dense layers?",
        default = 7)

parser.add_argument(
        "--minimum_final_dense_layers",
        type=int,
        help="Lowest number of neurons to try for final Dense layers, after "
             "the last ResidualMLP block, before the very last Dense layer "
             "returning an output? "
             "(0 doesn't create a layer')",
        default = 0)

parser.add_argument(
        "--maximum_final_dense_layers",
        type=int,
        help="Highest number of neurons to try for final Dense layers, after "
             "the last ResidualMLP block, before the very last Dense layer "
             "returning an output? "
             "(0 doesn't create a layer')",
        default = 150)

parser.add_argument(
        "--n_options_final_dense_layers",
        type=int,
        help="How many options for neurons to try for final Dense layers, "
             "after the last ResidualMLP block, before the very last "
             "Dense layer returning an output?",
        default = 7)



args, _ = parser.parse_known_args()
hparams = args.__dict__

# Boilerplate args

DATE = pendulum.now().__str__()[:16].replace("T","_").replace(":","_")
PROJECT_NAME = hparams["project_name"]
TRAINING_SET_SIZE = hparams["training_set_size"]

# Keras tuner search & fit args

PATIENCE = hparams["patience"]
PATIENCE_MIN_DELTA = hparams["patience_min_delta"]
BATCH_SIZE = hparams["batch_size"]
MAX_EPOCHS = hparams["max_epochs"]
RESULTS_DIR_FOR_SEARCH =\
    f'{DATE}_{PROJECT_NAME}_SEARCH_RUN'


# Base model args

BASE_MODEL_INPUT_SHAPE = (600,600,3)

# ResidualMLP model args

MINIMUM_LEARNING_RATE = hparams["minimum_learning_rate"]
MAXIMUM_LEARNING_RATE = hparams["maximum_learning_rate"]
NUMBER_OF_LEARNING_RATES_TO_TRY = hparams["number_of_learning_rates_to_try"]
MINIMUM_NUMBER_OF_BLOCKS = hparams["minimum_number_of_blocks"]
MAXIMUM_NUMBER_OF_BLOCKS = hparams["maximum_number_of_blocks"]
MINIMUM_NUMBER_OF_LAYERS_PER_BLOCK =\
    hparams["minimum_number_of_layers_per_block"]
MAXIMUM_NUMBER_OF_LAYERS_PER_BLOCK =\
    hparams["maximum_number_of_layers_per_block"]
MINIMUM_NEURONS_PER_BLOCK_LAYER = hparams["minimum_neurons_per_block_layer"]
MAXIMUM_NEURONS_PER_BLOCK_LAYER = hparams["maximum_neurons_per_block_layer"]
N_OPTIONS_OF_NEURONS_PER_LAYER_TO_TRY =\
    hparams["n_options_of_neurons_per_layer_to_try"]
MINIMUM_NEURONS_PER_BLOCK_LAYER_DECAY =\
    hparams["minimum_neurons_per_block_layer_decay"]
MAXIMUM_NEURONS_PER_BLOCK_LAYER_DECAY =\
    hparams["maximum_neurons_per_block_layer_decay"]
MINIMUM_DROPOUT_RATE_FOR_BYPASS_LAYERS =\
    hparams["minimum_dropout_rate_for_bypass_layers"]
MAXIMIM_DROPOUT_RATE_FOR_BYPASS_LAYERS =\
    hparams["maximim_dropout_rate_for_bypass_layers"]
N_OPTIONS_DROPOUT_RATE_FOR_BYPASS_LAYERS =\
    hparams["n_options_dropout_rate_for_bypass_layers"]
MINIMUM_INTER_BLOCK_LAYERS_PER_BLOCK =\
    hparams["minimum_inter_block_layers_per_block"]
MAXIMUM_INTER_BLOCK_LAYERS_PER_BLOCK =\
    hparams["maximum_inter_block_layers_per_block"]
N_OPTIONS_INTER_BLOCK_LAYERS_PER_BLOCK =\
    hparams["n_options_inter_block_layers_per_block"]
MINIMUM_DROPOUT_RATE = hparams["minimum_dropout_rate"]
MAXIMUM_DROPOUT_RATE = hparams["maximum_dropout_rate"]
N_OPTIONS_DROPOUT_RATE = hparams["n_options_dropout_rate"]
MINIMUM_FINAL_DENSE_LAYERS = hparams["minimum_final_dense_layers"]
MAXIMUM_FINAL_DENSE_LAYERS = hparams["maximum_final_dense_layers"]
N_OPTIONS_FINAL_DENSE_LAYERS = hparams["n_options_final_dense_layers"]

In [9]:
print(f"Date: {DATE}")
print(f"project_name: {PROJECT_NAME}")
print(f"training_set_size: {TRAINING_SET_SIZE}")
print(f"patience: {PATIENCE}")
print(f"PATIENCE_MIN_DELTA: {PATIENCE_MIN_DELTA}")
print(f"BATCH_SIZE: {BATCH_SIZE}")
print(f"MAX_EPOCHS: {MAX_EPOCHS}")
print(f"RESULTS_DIR_FOR_SEARCH: {RESULTS_DIR_FOR_SEARCH}")
print(f"MINIMUM_LEARNING_RATE {MINIMUM_LEARNING_RATE}")
print(f"MAXIMUM_LEARNING_RATE: {MAXIMUM_LEARNING_RATE}")
print(f"NUMBER_OF_LEARNING_RATES_TO_TRY: {NUMBER_OF_LEARNING_RATES_TO_TRY}")
print(f"MINIMUM_NUMBER_OF_BLOCKS: {MINIMUM_NUMBER_OF_BLOCKS}")
print(f"MAXIMUM_NUMBER_OF_BLOCKS: {MAXIMUM_NUMBER_OF_BLOCKS}")
print("MINIMUM_NUMBER_OF_LAYERS_PER_BLOCK: "
      f"{MINIMUM_NUMBER_OF_LAYERS_PER_BLOCK}")
print("MAXIMUM_NUMBER_OF_LAYERS_PER_BLOCK: "
      f"{MAXIMUM_NUMBER_OF_LAYERS_PER_BLOCK}")
print(f"MINIMUM_NEURONS_PER_BLOCK_LAYER: {MINIMUM_NEURONS_PER_BLOCK_LAYER}")
print(f"MAXIMUM_NEURONS_PER_BLOCK_LAYER: {MAXIMUM_NEURONS_PER_BLOCK_LAYER}")
print("N_OPTIONS_OF_NEURONS_PER_LAYER_TO_TRY: "
      f"{N_OPTIONS_OF_NEURONS_PER_LAYER_TO_TRY}")
print("MINIMUM_NEURONS_PER_BLOCK_LAYER_DECAY: "
      f"{MINIMUM_NEURONS_PER_BLOCK_LAYER_DECAY}")
print("MAXIMUM_NEURONS_PER_BLOCK_LAYER_DECAY: "
      f"{MAXIMUM_NEURONS_PER_BLOCK_LAYER_DECAY}")
print("MINIMUM_DROPOUT_RATE_FOR_BYPASS_LAYERS: "
      f"{MINIMUM_DROPOUT_RATE_FOR_BYPASS_LAYERS}")
print("MAXIMIM_DROPOUT_RATE_FOR_BYPASS_LAYERS: "
      f"{MAXIMIM_DROPOUT_RATE_FOR_BYPASS_LAYERS}")
print("N_OPTIONS_DROPOUT_RATE_FOR_BYPASS_LAYERS: "
      f"{N_OPTIONS_DROPOUT_RATE_FOR_BYPASS_LAYERS}")
print("MINIMUM_INTER_BLOCK_LAYERS_PER_BLOCK: "
      f"{MINIMUM_INTER_BLOCK_LAYERS_PER_BLOCK}")
print("MAXIMUM_INTER_BLOCK_LAYERS_PER_BLOCK: "
      f"{MAXIMUM_INTER_BLOCK_LAYERS_PER_BLOCK}")
print("N_OPTIONS_INTER_BLOCK_LAYERS_PER_BLOCK: "
      f"{N_OPTIONS_INTER_BLOCK_LAYERS_PER_BLOCK}")
print(f"MINIMUM_DROPOUT_RATE: {MINIMUM_DROPOUT_RATE}")
print(f"MAXIMUM_DROPOUT_RATE: {MAXIMUM_DROPOUT_RATE}")
print(f"N_OPTIONS_DROPOUT_RATE: {N_OPTIONS_DROPOUT_RATE}")
print(f"MINIMUM_FINAL_DENSE_LAYERS: {MINIMUM_FINAL_DENSE_LAYERS}")
print(f"MAXIMUM_FINAL_DENSE_LAYERS: {MAXIMUM_FINAL_DENSE_LAYERS}")
print(f"N_OPTIONS_FINAL_DENSE_LAYERS: {N_OPTIONS_FINAL_DENSE_LAYERS}")

Date: 2022-02-05_06_30
project_name: CIFAR10_EfficientNetB7-ResidualMLP_NAS_AUGMENTED_SPACE
training_set_size: 50000
patience: 25
PATIENCE_MIN_DELTA: 1e-05
BATCH_SIZE: 50
MAX_EPOCHS: 50
RESULTS_DIR_FOR_SEARCH: 2022-02-05_06_30_CIFAR10_EfficientNetB7-ResidualMLP_NAS_AUGMENTED_SPACE_SEARCH_RUN
MINIMUM_LEARNING_RATE 7e-05
MAXIMUM_LEARNING_RATE: 0.7
NUMBER_OF_LEARNING_RATES_TO_TRY: 7
MINIMUM_NUMBER_OF_BLOCKS: 1
MAXIMUM_NUMBER_OF_BLOCKS: 8
MINIMUM_NUMBER_OF_LAYERS_PER_BLOCK: 1
MAXIMUM_NUMBER_OF_LAYERS_PER_BLOCK: 8
MINIMUM_NEURONS_PER_BLOCK_LAYER: 30
MAXIMUM_NEURONS_PER_BLOCK_LAYER: 130
N_OPTIONS_OF_NEURONS_PER_LAYER_TO_TRY: 7
MINIMUM_NEURONS_PER_BLOCK_LAYER_DECAY: 0
MAXIMUM_NEURONS_PER_BLOCK_LAYER_DECAY: 50
MINIMUM_DROPOUT_RATE_FOR_BYPASS_LAYERS: 0.01
MAXIMIM_DROPOUT_RATE_FOR_BYPASS_LAYERS: 0.7
N_OPTIONS_DROPOUT_RATE_FOR_BYPASS_LAYERS: 7
MINIMUM_INTER_BLOCK_LAYERS_PER_BLOCK: 0
MAXIMUM_INTER_BLOCK_LAYERS_PER_BLOCK: 150
N_OPTIONS_INTER_BLOCK_LAYERS_PER_BLOCK: 7
MINIMUM_DROPOUT_RATE: 0.01
MAXI

### We instantiate an EfficientNetB7 previously trained on ImageNet, remove the final Dense and BatchNormalization layers after the last Conv2d layer. We set the last Conv2D layer to trainable.  We pass the output of this to a ResidualMLP model of variable hyperparameterized neural architecture. We set the hyperparameter and neural architecture search space to be explored.

In [10]:
mod_with_fc_raw = tf.keras.applications.efficientnet.EfficientNetB7(
    include_top=True, weights='imagenet', input_tensor=None,
    input_shape = BASE_MODEL_INPUT_SHAPE, pooling='max', classes=1000
)

# Make the deepest conv2d layer trainable, leave everything else
# as not trainable
for layer in mod_with_fc_raw.layers:
    layer.trainable = False
# Last conv2d layer. This we want to train .
mod_with_fc_raw.layers[-6].trainable = True

# Create the final base model
# (remove the final Dense and BatchNormalization layers ...) 
efficient_net_b_7_transferable_base_model =\
    tf.keras.Model(inputs=mod_with_fc_raw.layers[0].input, 
                    outputs=mod_with_fc_raw.layers[-3].output)

model_builder = ResidualMLP(
                    problem_type= "classification",  
                    minimum_learning_rate = MINIMUM_LEARNING_RATE, 
                    maximum_learning_rate = MAXIMUM_LEARNING_RATE, 
                    number_of_learning_rates_to_try =
                        NUMBER_OF_LEARNING_RATES_TO_TRY, 
                    input_shape = (32,32,3), 
                    bw_images = False, 
                    base_model = 
                        efficient_net_b_7_transferable_base_model, 
                    base_model_input_shape = BASE_MODEL_INPUT_SHAPE, 
                    flatten_after_base_model = False, 
                    minimum_number_of_blocks = MINIMUM_NUMBER_OF_BLOCKS, 
                    maximum_number_of_blocks = MAXIMUM_NUMBER_OF_BLOCKS, 
                    minimum_number_of_layers_per_block =
                        MINIMUM_NUMBER_OF_LAYERS_PER_BLOCK, 
                    maximum_number_of_layers_per_block =
                        MAXIMUM_NUMBER_OF_LAYERS_PER_BLOCK,
                    minimum_neurons_per_block_layer =
                        MINIMUM_NEURONS_PER_BLOCK_LAYER, 
                    maximum_neurons_per_block_layer =
                        MAXIMUM_NEURONS_PER_BLOCK_LAYER, 
                    n_options_of_neurons_per_layer_to_try =
                        N_OPTIONS_OF_NEURONS_PER_LAYER_TO_TRY, 
                    minimum_neurons_per_block_layer_decay =
                        MINIMUM_NEURONS_PER_BLOCK_LAYER_DECAY, 
                    maximum_neurons_per_block_layer_decay = 
                        MAXIMUM_NEURONS_PER_BLOCK_LAYER_DECAY, 
                    minimum_dropout_rate_for_bypass_layers =
                        MINIMUM_DROPOUT_RATE_FOR_BYPASS_LAYERS, 
                    maximim_dropout_rate_for_bypass_layers =
                        MAXIMIM_DROPOUT_RATE_FOR_BYPASS_LAYERS, 
                    n_options_dropout_rate_for_bypass_layers =
                        N_OPTIONS_DROPOUT_RATE_FOR_BYPASS_LAYERS,
                    minimum_inter_block_layers_per_block =
                        MINIMUM_INTER_BLOCK_LAYERS_PER_BLOCK, 
                    maximum_inter_block_layers_per_block =
                        MAXIMUM_INTER_BLOCK_LAYERS_PER_BLOCK,
                    n_options_inter_block_layers_per_block =\
                        N_OPTIONS_INTER_BLOCK_LAYERS_PER_BLOCK,
                    minimum_dropout_rate = MINIMUM_DROPOUT_RATE, 
                    maximum_dropout_rate = MAXIMUM_DROPOUT_RATE,
                    n_options_dropout_rate = N_OPTIONS_DROPOUT_RATE, 
                    minimum_final_dense_layers =
                        MINIMUM_FINAL_DENSE_LAYERS,
                    maximum_final_dense_layers =
                        MAXIMUM_FINAL_DENSE_LAYERS, 
                    n_options_final_dense_layers =
                        N_OPTIONS_FINAL_DENSE_LAYERS, 
                    number_of_classes = 10,
                    final_activation = tf.keras.activations.softmax)

268337152/268326632 [==============================] - 12s 0us/step


## We restore the hyperband tunerused in task.py and view the summaru of results:

In [11]:

DATE = pendulum.now().__str__()[:16].replace("T","_").replace(":","_")



RESULTS_DIR_FOR_SEARCH =\
    f'{DATE}_{PROJECT_NAME}_SEARCH_RUN'

logdir_for_search = os.path.join("logs", RESULTS_DIR_FOR_SEARCH + "_TB")
tensorboard_callback_search =\
    tf.keras.callbacks.TensorBoard(logdir_for_search, histogram_freq=1)

tuner = kt.Hyperband(
    model_builder.build_auto_residual_mlp,
    objective='val_loss',
    max_epochs = MAX_EPOCHS,
    hyperband_iterations = 2)

INFO:tensorflow:Reloading Oracle from existing project ./untitled_project/oracle.json


2022-02-05 06:30:27.003881: I tensorflow/core/profiler/lib/profiler_session.cc:131] Profiler session initializing.
2022-02-05 06:30:27.003932: I tensorflow/core/profiler/lib/profiler_session.cc:146] Profiler session started.
2022-02-05 06:30:27.003973: I tensorflow/core/profiler/internal/gpu/cupti_tracer.cc:1614] Profiler found 1 GPUs
2022-02-05 06:30:27.094328: I tensorflow/core/profiler/lib/profiler_session.cc:164] Profiler session tear down.
2022-02-05 06:30:27.094538: I tensorflow/core/profiler/internal/gpu/cupti_tracer.cc:1749] CUPTI activity buffer flushed


All permutations:
       number_of_blocks  layers_per_block  neurons_per_block_layer  \
0                     1                 1                       30   
1                     1                 1                       30   
2                     1                 1                       30   
3                     1                 1                       30   
4                     1                 1                       30   
...                 ...               ...                      ...   
22843                 8                 8                      130   
22844                 8                 8                      130   
22845                 8                 8                      130   
22846                 8                 8                      130   
22847                 8                 8                      130   

       neurons_per_block_layer_decay  
0                                  0  
1                                  1  
2                       

In [12]:
tuner.results_summary(10)

Results summary
Results in ./untitled_project
Showing 10 best trials
Objective(name='val_loss', direction='min')
Trial summary
Hyperparameters:
learning_rate: 0.009343337171789316
blocks: 399
bypass_layers_units: 46
inter_block_layers: 25
final_dense_layers: 150
b_norm_or_dropout_residual_bypass_layers: dropout
dropout_rate_for_bypass_layers: 0.24
b_norm_or_dropout_last_layers: bnorm
dropout_rate: 0.355
tuner/epochs: 2
tuner/initial_epoch: 0
tuner/bracket: 3
tuner/round: 0
Score: 0.3171733617782593
Trial summary
Hyperparameters:
learning_rate: 0.0004635558198557901
blocks: 1702
bypass_layers_units: 30
inter_block_layers: 25
final_dense_layers: 25
b_norm_or_dropout_residual_bypass_layers: dropout
dropout_rate_for_bypass_layers: 0.24
b_norm_or_dropout_last_layers: bnorm
dropout_rate: 0.12499999999999999
tuner/epochs: 2
tuner/initial_epoch: 0
tuner/bracket: 3
tuner/round: 0
Score: 0.328523188829422
Trial summary
Hyperparameters:
learning_rate: 0.03967702188815936
blocks: 1747
bypass_layer

# Best architecture so far (12 hours of searching on a node with 2 A6000 GPUs and 90 GB RAM): 
### Residual blocks:  [[2, 63, 28]]
1. This means: The Input() layer's output proceeds to on Dense(63) layer.
2. The output of this is duplicated into 2 tensors: x and y.
3. x proceeds through a block of 2 Dense layers:
    1. Dense(63) - then BatchNormalization 
    2. Dense(33) - then BatchNormalization 
4. y bypasses this block and proceeds through: 
    1. One Dense(46) layer in the bypass path followed by 
    2. Dropout(0.24)
4. The outputs of x and y merge in a concat layer.

### The output of this goes through:
1. One Dense(150) 
2. a batch notmalization layer
3. Dense(10) with softmax. (model's output)

### Note that in this architecture, inter_block_layers is ignored, as this only applies to architectures consisting of more than one residual block.

Val loss: 0.317


## Let's examine the best model:

In [13]:
best_model_so_far_2022_02_04 = tuner.get_best_models(num_models=1)[0]

All permutations:
       number_of_blocks  layers_per_block  neurons_per_block_layer  \
0                     1                 1                       30   
1                     1                 1                       30   
2                     1                 1                       30   
3                     1                 1                       30   
4                     1                 1                       30   
...                 ...               ...                      ...   
22843                 8                 8                      130   
22844                 8                 8                      130   
22845                 8                 8                      130   
22846                 8                 8                      130   
22847                 8                 8                      130   

       neurons_per_block_layer_decay  
0                                  0  
1                                  1  
2                       

### Let's look at the details of the model:

In [14]:
best_model_so_far_2022_02_04.summary()

Model: "model"
__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param #     Connected to                     
input_1 (InputLayer)            [(None, 32, 32, 3)]  0                                            
__________________________________________________________________________________________________
resizing (Resizing)             (None, 600, 600, 3)  0           input_1[0][0]                    
__________________________________________________________________________________________________
model (Functional)              (None, 2560)         64097687    resizing[0][0]                   
__________________________________________________________________________________________________
dense (Dense)                   (None, 63)           161343      model[1][0]                      
______________________________________________________________________________________________

### Let's see how this performs on the test set:

In [15]:
best_model_so_far_2022_02_04.evaluate(x_test,y_test_ohe)

2022-02-05 06:30:51.710845: I tensorflow/compiler/mlir/mlir_graph_optimization_pass.cc:185] None of the MLIR Optimization Passes are enabled (registered 2)
2022-02-05 06:30:56.204242: I tensorflow/stream_executor/cuda/cuda_dnn.cc:381] Loaded cuDNN version 8204


  1/313 [..............................] - ETA: 45:00 - loss: 0.3043 - top_1_categorical_accuracy: 0.9375 - top_2_categorical_accuracy: 0.9375 - top_3_categorical_accuracy: 0.9375 - top_4_categorical_accuracy: 0.9688 - top_5_categorical_accuracy: 0.9688 - top_6_categorical_accuracy: 0.9688 - top_7_categorical_accuracy: 0.9688 - top_8_categorical_accuracy: 1.0000 - top_9_categorical_accuracy: 1.0000 - precision: 0.9677 - recall: 0.9375 - accuracy: 0.0000e+00

2022-02-05 06:31:00.354400: I tensorflow/stream_executor/cuda/cuda_blas.cc:1760] TensorFloat-32 will be used for the matrix multiplication. This will only be logged once.


313/313 [==============================] - 422s 1s/step - loss: 0.3247 - top_1_categorical_accuracy: 0.8986 - top_2_categorical_accuracy: 0.9608 - top_3_categorical_accuracy: 0.9792 - top_4_categorical_accuracy: 0.9888 - top_5_categorical_accuracy: 0.9931 - top_6_categorical_accuracy: 0.9962 - top_7_categorical_accuracy: 0.9973 - top_8_categorical_accuracy: 0.9989 - top_9_categorical_accuracy: 0.9996 - precision: 0.9294 - recall: 0.8670 - accuracy: 1.9000e-04


[0.32468482851982117,
 0.8985999822616577,
 0.9607999920845032,
 0.979200005531311,
 0.9887999892234802,
 0.9930999875068665,
 0.9962000250816345,
 0.9973000288009644,
 0.9988999962806702,
 0.9995999932289124,
 0.929360032081604,
 0.8669999837875366,
 0.0001900000061141327]

## Results:
### Test set top-1 categorical accuracy: 0.8985999822616577

# Conclusion and next steps

 This is a good result, however, it appears I need to keep working and develop an algorithym that does a reasonable 1st pass broad search to narrow down the vast neural architecture and hyperparameter search space (like a random or grid search of a small number of evenly distributed samples of the search space to find hotspots and cold regions in the space), then automatically set up a  hyperband focusing on the spaces in and around these hotspots.  It does appear that a cold - start hyperband search usually takes 50 - 150 trials to find an optimum in the broad search space with a smaller dataset.This works well and is efficient, but it may be best to eliminate areas of the search space That are unlikely to hold an optimal neural architecture and limit the hyperband to spaces found on a computationally efficient first pass search to be likely to hold an optimum.